File: 2023/02. Cube Conundrum/c++/one/src/main.cpp
#include <algorithm>
#include <array>
#include <cstring>
#include <fstream>
#include <cstddef>
#include <iomanip>
#include <iostream>
#include <iterator>
#include <string>
#include <string_view>
#include <vector>

static constexpr std::string_view DELIMS{" :;,"};
static const std::size_t NUM_COLORS{3};

static constexpr std::array<std::string_view, NUM_COLORS> COLOR_STRINGS{
    "red",
    "green",
    "blue"
};

static constexpr std::array<int, NUM_COLORS> MAX_CUBES{
    12, // red
    13, // green
    14  // blue
};

int main(const int argc, const char** const argv)
{
    if (argc != 2)
    {
        std::cerr << "Usage: " << argv[0] << " <INPUT_FILE\n>";
        return 1;
    }

    std::ifstream infile{argv[1]};
    if (!infile.is_open())
    {
        std::cerr << "ERROR! Could not open input file\n";
        return 1;
    }

    // Puzzle result
    int idSum{};

    // Read each line
    int lineNum{1};
    for (std::string line; std::getline(infile, line); ++lineNum)
    {
        // Minimum number of cubes of each color needed for current game to be possilbe
        std::array<int, NUM_COLORS> minNeeded{};

        // Copy line input mutable buffer for std::strtok
        std::vector<char> lineBuffer(line.size() + 1);
        std::copy(line.begin(), line.end(), lineBuffer.begin());

        // Skip first two tokens
        std::strtok(lineBuffer.data(), DELIMS.data());
        std::strtok(nullptr, DELIMS.data());

        // Read ammount, color pairs
        for (const char* ammountString{std::strtok(nullptr, DELIMS.data())}; ammountString != nullptr; ammountString = std::strtok(nullptr, DELIMS.data()))
        {
            const int ammount{std::stoi(ammountString)};
            const std::string_view colorString{std::strtok(nullptr, DELIMS.data())};

            // Get index for color string or print erorr message if color is not found
            const auto colorIt{std::ranges::find(COLOR_STRINGS, colorString)};
            if (colorIt == COLOR_STRINGS.end())
            {
                std::cerr << "Unknown color " << std::quoted(colorString) << " on line " << lineNum << '\n';
                return 1;
            }
            else
            {
                // If ammount is greater than current max for given color record new max
                const std::size_t colorIndex{static_cast<std::size_t>(std::distance(COLOR_STRINGS.begin(), colorIt))};
                minNeeded[colorIndex] = std::max(minNeeded[colorIndex], ammount);
            }
        }

        // Check if game is possible in regards to all 3 colors
        bool possible{true};
        for (std::size_t i{}; i < NUM_COLORS; ++i)
        {
            if (minNeeded[i] > MAX_CUBES[i])
            {
                possible = false;
                break;
            }
        }

        // Add lineId to total if game is possible
        if (possible)
            idSum += lineNum;
    }

    std::cout << "Result: " << idSum << '\n';

    return 0;
}

File: 2023/02. Cube Conundrum/rust/one/src/main.rs
use std::env;
use std::fs;
use std::iter::zip;

const NUM_COLORS: usize = 3;
const MAX_CUBES: [i32; NUM_COLORS] = [
    12, // red
    13, // green
    14, // blue
];
const COLOR_STRINGS: [&str; NUM_COLORS] = ["red", "green", "blue"];

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <INPUT_FILE>", args[0]);
        return;
    }

    // Puzzle result
    let mut id_sum = 0;

    // Read line by line
    let contents = fs::read_to_string(&args[1]).expect("Could not read input file");
    for (line_num, line) in contents.lines().enumerate() {
        // Min number of cubes of each color needed for game to be possible
        let mut min_needed = [0; NUM_COLORS];

        // Sperate line into tokens, skip the "Game" and game id tokens, and remove emtpy spaces
        let mut tokens = line
            .split(|c| " :;,".contains(c))
            .skip(2)
            .filter(|part| !part.is_empty());

        // Iterate over each ammount, color pair
        while let Some(ammount) = tokens.next() {
            // Parse the ammount
            let ammount: i32 = ammount
                .parse()
                .expect(&format!("Could not convert \"{ammount}\" to integer"));

            // Parse the color
            let color = tokens
                .next()
                .expect(&format!("Expected color on line {line_num}"));

            // Get the 0-2 index for the color
            let color = COLOR_STRINGS
                .iter()
                .enumerate()
                .find(|(_, color_string)| **color_string == color)
                .map(|(index, _)| index)
                .expect(&format!("Unknown color \"{color}\" on line {line_num}"));

                // Update min needed of ammount is greater
                min_needed[color] = min_needed[color].max(ammount);
        }

        // Check of game is possible
        let mut possible = true;
        for (needed, max) in zip(&min_needed, &MAX_CUBES) {
            if needed > max {
                possible = false;
            }
        }

        // Add game ID to total sum if game is possible
        if possible {
            id_sum += line_num + 1;
        }
    }

    println!("Result: {id_sum}");
}
